src/models/bundle-descriptor-constraints.ts

```diff
@@ -209,179 +214,239 @@ const MICROSERVICE_CONSTRAINTS: ObjectConstraints<Microservice> = {
   }
 }
 
-const MICROFRONTEND_CONSTRAINTS: UnionTypeConstraints<MicroFrontend> = {
-  constraints: [
-    {
-      name: {
-        required: true,
-        type: 'string',
-        validators: [nameRegExpValidator]
-      },
-      stack: {
-        required: true,
-        type: 'string',
-        validators: [values(MicroFrontendStack)]
-      },
-      type: {
-        required: true,
-        type: 'string',
-        validators: [
-          values([MicroFrontendType.Widget, MicroFrontendType.WidgetConfig])
-        ]
-      },
-      titles: {
-        required: true,
-        validators: [isMapOfStrings],
-        children: {}
-      },
-      publicFolder: {
+const WIDGET_MICROFRONTEND_CONSTRAINTS: ObjectConstraints<WidgetMicroFrontend> = {
+  name: {
+    required: true,
+    type: 'string',
+    validators: [nameRegExpValidator]
+  },
+  stack: {
+    required: true,
+    type: 'string',
+    validators: [values(MicroFrontendStack)]
+  },
+  titles: {
+    required: true,
+    validators: [isMapOfStrings],
+    children: {}
+  },
+  publicFolder: {
+    required: false,
+    type: 'string'
+  },
+  group: {
+    required: true,
+    type: 'string'
+  },
+  apiClaims: {
+    isArray: true,
+    required: false,
+    children: API_CLAIMS_CONSTRAINTS
+  },
+  nav: {
+    isArray: true,
+    required: false,
+    children: NAV_CONSTRAINTS
+  },
+  commands: {
+    required: false,
+    children: {
+      build: {
         required: false,
         type: 'string'
-      },
-      group: {
-        required: true,
-        type: 'string'
-      },
-      apiClaims: {
-        isArray: true,
-        required: false,
-        children: API_CLAIMS_CONSTRAINTS
-      },
-      nav: {
-        isArray: true,
-        required: false,
-        children: NAV_CONSTRAINTS
-      },
-      commands: {
-        required: false,
-        children: {
-          build: {
-            required: false,
-            type: 'string'
-          }
-        }
       }
-    },
-    {
-      name: {
-        required: true,
-        type: 'string',
-        validators: [nameRegExpValidator]
-      },
-      stack: {
-        required: true,
-        type: 'string',
-        validators: [values(MicroFrontendStack)]
-      },
-      titles: {
-        required: true,
-        validators: [isMapOfStrings],
-        children: {}
-      },
-      publicFolder: {
+    }
+  },
+  type: {
+    required: true,
+    type: 'string',
+    validators: [values(MicroFrontendType)]
+  },
+  contextParams: {
+    isArray: true,
+    required: false,
+    type: 'string',
+    validators: [values(WidgetContextParam)]
+  }
+}
+
+const WIDGETCONFIG_MICROFRONTEND_CONSTRAINTS: ObjectConstraints<WidgetConfigMicroFrontend> = {
+  name: {
```

Comment:
Can't we group all constrainst common to all MFE types (name, type, ..) without repeating them?

Reply:
It's quite tricky to do that due to how the `ObjectConstraints` type is currently structured. It can probably be done via something like:
```
type BaseMicroFrontendConstraints = ObjectConstraints<BaseMicroFrontendConstraints> // the type to use for the common properties
type WidgetMicroFrontendConstraints = BaseMicroFrontendConstraints & ObjectConstraints<WidgetMicroFrontend>
...
```
 Any thoughts @zonia3000?

Reply:
Ok let's stay with this if it's too much complicated